internal/appsec/config/wafmanager.go

```diff
@@ -0,0 +1,206 @@
+// version of the rules that were used to build it.
+func (m *WAFManager) NewHandle() (*libddwaf.Handle, string) {
+	m.mu.RLock()
+	rulesVersion := m.rulesVersion
+	hdl := m.builder.Build()
+	m.mu.RUnlock()
+	return hdl, rulesVersion
+}
+
+// Close releases all resources associated with this [WAFManager].
+func (m *WAFManager) Close() {
+	m.doClose(false)
+}
+
+func (m *WAFManager) doClose(leaked bool) {
+	m.mu.Lock()
+	if m.closed {
+		m.mu.Unlock()
+		return
+	}
+	if leaked {
+		log.Warn("WAFManager was leaked and is being closed by GC. Remember to call WAFManager.Close() explicitly!")
+	}
+
+	m.builder.Close()
+	m.rulesVersion = ""
+	m.closed = true
+	m.mu.Unlock()
+}
+
+// RemoveConfig removes a configuration from the receiving [WAFManager].
+func (m *WAFManagerUpdater) RemoveConfig(path string) {
+	m.manager.builder.RemoveConfig(path)
+}
+
+// RemoveDefaultConfig removes the initial configuration from the receiving [WAFManager]. Returns
+// true if the default config was actually removed; false otherwise (e.g, if it had previously been
+// removed, or there was no default config to begin with).
+func (m *WAFManagerUpdater) RemoveDefaultConfig() bool {
+	removed := false
+	if m.manager.defaultLoaded {
+		removed = m.manager.builder.RemoveConfig(defaultRulesPath)
+		m.manager.defaultLoaded = false
+	}
+	return removed
+}
+
+// AddOrUpdateConfig adds or updates a configuration in the receiving [WAFManager].
+func (m *WAFManagerUpdater) AddOrUpdateConfig(path string, fragment any) (libddwaf.Diagnostics, error) {
+	diag, err := m.manager.builder.AddOrUpdateConfig(path, fragment)
+	if err == nil && diag.Version != "" {
+		m.manager.rulesVersion = diag.Version
+	}
+
+	// Submit the telemetry metrics for error counts obtained from the [libddwaf.Diagnostics] object.
+	// See: https://docs.google.com/document/d/1lcCvURsWTS_p01-MvrI6SmDB309L1e8bx9txuUR1zCk/edit?tab=t.0#heading=h.nwzm8andnx41
+	eventRulesVersion := diag.Version
+	if eventRulesVersion == "" {
+		eventRulesVersion = m.manager.rulesVersion
+	}
+	diag.EachFeature(func(name string, feat *libddwaf.Feature) {
+		errCount := telemetry.Count(telemetry.NamespaceAppSec, "waf.config_errors", []string{
+			"waf_version:" + libddwaf.Version(),
+			"event_rules_version:" + eventRulesVersion,
+			"config_key:" + name,
+			"scope:item",
+			"action:update",
+		})
+		errCount.Submit(0)
+		for _, ids := range feat.Errors {
+			errCount.Submit(float64(len(ids)))
+		}
+	})
+
+	return diag, err
+}
+
+// RestoreDefaultConfig restores the initial configurations to the receiving [WAFManager].
+func (m *WAFManagerUpdater) RestoreDefaultConfig() error {
+	if m.manager.initRules == nil {
+		return nil
+	}
+	_, err := m.AddOrUpdateConfig(defaultRulesPath, m.manager.initRules)
+	return err
+}
+
+// Unlock releases the lock on the [WAFManager]. The receiving [WAFManagerUpdater] must not be
+// used after calling this method.
+func (m *WAFManagerUpdater) Unlock() {
+	m.manager.mu.Unlock()
+	m.manager = nil
```

Comment:
```suggestion
	m.manager = nil
	m.manager.mu.Unlock()
```

There is a data race here if we don't reverse the statements no ? :thinking:

Reply:
If you nil-out the manager before unlocking it... you'll hit an NPE...

And the data race exists... only if you're mis-using the WAFManager anyway -- this is the `Unlock` function, akin to `Close` -- if you race here that means you're trying to use-after-Unlock which is illegal.